Proposal for strong decoupling between boilerplate and app that uses it

MikeTschudi opened the pull request:
#85

driskull commented on config/config.json:
```diff
@@ -1,7 +1,7 @@
   "appid": "",
   "group": "",
-  "webscene": "19faa71a3bf6468cae35b4fce9393a7d",
+  "webscene": "",
```
@MikeTschudi why are we removing the default webscene?

MikeTschudi replied:
It was a tradeoff to permit running the boilerplate with ?webmap=*id*, ?webscene=*id*, ?group=*id* out of the box.  With the webscene id in config.json, running ?group=*id* brings up that webscene rather than the group.  It means that the boilerplate won't bring up a default display if no URL parameters are provided, but that seems more intuitive than having ?group=*id* bring up a webscene.

driskull replied:
I guess. But when someone loads the app without any URL param its going to fail. :<

kellyhutchins replied:
So should we require a url param and if one isn't there then the error page displays instead of the default scene?

MikeTschudi replied:
But that's much more an "ok-that-makes-sense" behavior than to supply a group id and get a web scene.

driskull replied:
I'm ok with it if @kellyhutchins is.

MikeTschudi replied:
(Overlapping comments) @kellyhutchins:  If no url param is supplied, one sees the error page with "main:: Could not load an item to display"

kellyhutchins replied:
All makes sense to me. Let's leave out the defaults.

driskull commented on js/application/appInit.js:
```diff
@@ -0,0 +1,270 @@
+          document.title = this.config.title;
+
+          // Launch the app specifics
+          this.main.runMap(this.config, webscene, view);
```
@MikeTschudi should this be runScene?

MikeTschudi replied:
I agree that there might be a better name than `runMap`.  I was trying to distinguish it from `runGroup`.

Because of inheritance, the same function can be used for 2D and 3D because WebMap and WebScene both inherit from Map (second argument) and MapView and SceneView both inherit from View (third argument).  This permits the boilerplate to have just two `run*` functions.  But that might be a false economy--perhaps it is clearer to have a run for 2D, a run for 3D, and a run for groups?

driskull commented on js/dojoConfig.js:
```diff
@@ -5,7 +5,7 @@ dojoConfig = {
     name: "application",
     location: package_path + "/js/application",
-    main: "main"
+    main: "appInit"
```
lets just rename to "init" because its already in the application folder. doesn't need "appInit". @MikeTschudi

MikeTschudi replied:
Sounds like an improvement to me. Via a followup pull request?

driskull replied:
Can you do it in this one?

driskull commented:
@MikeTschudi what about instead of `runGroup`, `runMap`. We just have `start` that gets passed an object that has properties of all the necessary things. 

```
{
config:...,
view:...,
webscene:...,
webmap:...,
groupInfo:...,
groupItems:...
}
```

MikeTschudi commented:
Also sounds like an improvement to me.

driskull commented on js/application/init.js:
```diff
@@ -194,7 +194,7 @@ define([
 
           // Launch the app specifics
-          this.main.runMap(this.config, webmap, view);
+          this.main.start(this.config, dom.byId("viewDiv"), view, webmap, null, null);
```
@MikeTschudi can we make one object passed as an argument instead of multiple arguments with null values?

MikeTschudi replied:
Nice.  Better conformance with 4.x style.

driskull commented on js/application/init.js:
```diff
@@ -194,7 +194,12 @@ define([
+            config: this.config,
+            viewDiv: dom.byId("viewDiv"),
+            view: view,
+            mapOrScene: webmap
```
can we separate these out into `webmap` and `webscene` properties? Then users don't have to try to figure out which it is. and using it is easy as `object.webmap || object.webscene`

driskull commented on js/application/init.js:
```diff
@@ -194,7 +194,12 @@ define([
-          this.main.start(this.config, dom.byId("viewDiv"), view, webmap, null, null);
+          this.main.start({
+            config: this.config,
+            viewDiv: dom.byId("viewDiv"),
```
rename to `viewNode`.

driskull commented on js/application/init.js:
```diff
@@ -242,7 +247,12 @@ define([
+            config: this.config,
+            viewDiv: dom.byId("viewDiv"),
+            view: view,
+            mapOrScene: webscene
```
@MikeTschudi can we just get this renamed to `scene:...` or `map:...`? That way users know which it is.

MikeTschudi replied:
But they could also test `options.view instanceof MapView` and `options.mapOrScene instanceof WebMap` if their app happens to support both 2D and 3D.

driskull replied:
Yeah, but lets just tell them what it is upfront so they don't have to do that. There's really no reason to have it as a combined property that handles different things.

driskull commented on js/application/init.js:
```diff
@@ -0,0 +1,285 @@
+            config: this.config,
+            viewNode: dom.byId("viewDiv"),
+            view: view,
+            mapOrScene: webmap
```
@MikeTschudi it just seems very odd to have a property called mapOrScene. Just have the property be what it is.

MikeTschudi replied:
True. That's why it originally named after the superclass like the view property, but `map`, even though standing for the `Map` hierarchy, seemed unclear for `WebMap` or `WebScene`.

driskull replied:
yeah. Just make it separate properties webmap or webscene.

alaframboise commented:
@driskull @MikeTschudi Just started working with this yesterday. Integrating your app code and references is definitely tricky.

Question: Where do we recommend people to put their app code? main.js? To configure the app and page, some code needs to be applied to the document framework e.g. CSS styles, title, description... and other code needs to be applied to the view properties e.g. padding before creation. Right now I'm just putting everything in main.js. I was looking for an app.js...

And I'm still trying to figure out the initialization (below). I guess I expected to:

1. Create a boilerplate object().
2. Get a webItem back.
3. Do my app template stuff 
 a) Set document HTML/CSS and element 
 b) Load map or scene
 c) Other app specific stuff...

```
   require([
      "boilerplate",
      "dojo/text!boilerplate/settings.json",
      "application"
    ], function (
      Boilerplate,
      boilerplateSettings,
      applicationMain
    ) {

      // create my main application. Start placing your logic in the main.js file.
      var mainApp = new applicationMain();

      // create the template. This will take care of all the logic required for template applications. If unsuccessful, we're calling error on main. Otherwise, we call startup on main.
      new Boilerplate(JSON.parse(boilerplateSettings)).then(mainApp.init.bind(mainApp), mainApp.reportError.bind(mainApp));

    });
```

MikeTschudi commented:
@alaframboise:  The recommendation is to put the app code into main.js, with optional changes in index.html and a few other places.  It probably won't be necessary to modify the code in the snippet that you showed.

In this pull request, you'll see three entry points into main.js:

1. adjustConfig:  uncomment this if you want to modify the configuration before the webmap/webscene/group gallery is created
2. reportError:  uncomment this if you want to change the error display that the boilerplate uses
3. start:  put your app code in here

Here are the places where I've made changes in my copy of the boilerplate to support a 3D 4.0 app:

1. index.html
  * Added a link to my css/main.css after the other links
  * Added some divs after the other divs
  * Added script includes for my library routines after the other scripts
2. config\config.json
  * Defined app defaults
3. js\boilerplate\settings.json
  * Disabled fetches for webmaps and groups
  * Removed URL params not used in app
4. js\application\main.js
  * Completely changed but for start() signature
5. js\application\nls\resources et al.
  * Added strings
6. resources\configurationPanel.js
  * Defined hosted app parameters

Plus I've added app-specific files in various places.

The boilerplate files that I modified don't tend to change, so I've found it no trouble to merge boilerplate updates several times a day into my app as we refine the decoupling design.  For example, the boilerplate.js and init.js files have changed during the refinement, but because of the three-entry-point contract between the boilerplate and the main.js file, none of those changes required any changes to main.js for me--I just copied the files in.

driskull commented:
@MikeTschudi the only remaining thing before merging is to separate the mapOrScene property like this:

```
{
webscene:...,
webmap:...,
}
```

MikeTschudi commented:
Except that I prefer the polymorphism  :-)

alaframboise commented:
what about this

```
map: {
 url: "19faa71a3bf6468cae35b4fce9393a7d" 
 type: "webmap"
}
```

MikeTschudi commented:
What we're debating is the signature of the start() function in main.js. It contains a single argument `options` with properties appropriate to what the boilerplate finds in the configuration (i.e., webmap app or webscene app or group gallery app?)

Two proposals:
```
options: {
  config,     // configuration assembled by boilerplate
  viewNode,   // dom node for main app display
  view,       // resolved MapView or SceneView created in init
  map,        // resolved WebMap or WebScene created in init
  groupInfo,  // ArcGIS item groupInfo for group
  groupItems  // list of groupItems in group
}
```

```
options: {
  config,     // configuration assembled by boilerplate
  viewNode,   // dom node for main app display
  view,       // resolved MapView or SceneView created in init
  webScene,   // resolved WebScene created in init
  webMap,     // resolved WebMap created in init
  groupInfo,  // ArcGIS item groupInfo for group
  groupItems  // list of groupItems in group
}
```

Because of possible ambiguity, the `map` property has also been proposed to be `mapOrScene`, but that's a bit awkward. To be parallel with `view`, maybe it should be `web`.  :-)

kellyhutchins commented:
I prefer option 2 because in my opinion it makes it more clear what value is expected (web map or web scene) vs the map option which I think could be confusing.

MikeTschudi commented:
You broke the tie; I'll make the change!

kellyhutchins commented:
Thanks @MikeTschudi  for making these updates. I think your modifications will make it easier for template devs to keep in sync with any updates.

alaframboise commented:
I just merged and this helps keep the app logic in one location. Thanks!
